crates/common/chain/lean/src/lean_chain.rs

```diff
@@ -360,8 +379,11 @@ impl LeanChain {
     ///
     /// See lean specification:
     /// <https://github.com/leanEthereum/leanSpec/blob/ee16b19825a1f358b00a6fc2d7847be549daa03b/docs/client/forkchoice.md?plain=1#L314-L342>
-    pub async fn on_block(&mut self, signed_block: SignedBlock) -> anyhow::Result<()> {
-        let block_hash = signed_block.message.tree_hash_root();
+    pub async fn on_block(
+        &mut self,
+        signed_block: SignedBlockWithAttestation,
+    ) -> anyhow::Result<()> {
+        let block_hash = signed_block.message.block.tree_hash_root();
```

Comment:
```suggestion
        let block = signed_block.message.block;
        let block_hash = block.tree_hash_root();
```

This will make our life easier for the rest of this function body.

Reply:
```rust
    pub async fn on_block(
        &mut self,
        signed_block: SignedBlockWithAttestation,
    ) -> anyhow::Result<()> {
        let block = &signed_block.message.block;
        let block_hash = block.tree_hash_root();

        let (lean_block_provider, latest_justified_provider, lean_state_provider) = {
            let db = self.store.lock().await;
            (
                db.lean_block_provider(),
                db.latest_justified_provider(),
                db.lean_state_provider(),
            )
        };

        // If the block is already known, ignore it
        if lean_block_provider.contains_key(block_hash) {
            return Ok(());
        }

        let mut state = lean_state_provider.get(block.parent_root)?.ok_or_else(|| {
            anyhow!(
                "Parent state not found for block: {block_hash}, parent: {}",
                block.parent_root
            )
        })?;

        // TODO: Add signature validation once spec is complete.
        state.state_transition(&block, true, true)?;

        let mut signed_attestations = vec![];
        for attestation in &block.body.attestations {
            signed_attestations.push(SignedAttestation {
                message: attestation.clone(),
                signature: FixedBytes::<4000>::default(),
            });
        }
        lean_block_provider.insert(block_hash, signed_block)?;
        latest_justified_provider.insert(state.latest_justified)?;
        lean_state_provider.insert(block_hash, state)?;
        self.on_attestation_from_block(signed_attestations).await?;
        self.update_head().await?;

        Ok(())
    }
```

My bad, we might get the pointer from the `signed_block` like this.